src/frontends/src/query_frontend.hpp

```diff
@@ -301,26 +315,52 @@ class RequestHandler {
     }
 
     AppMetrics app_metrics(name);
+    
+    /*
+   * JSON format for prediction query request:
+   * {
+   *  "input" := [double] | [int] | [string] | [byte] | [float]
+   *  "input_batch" := [[double] | [int] | [byte] | [float] | string]
+   *  "version" := string (optional)
+   * }
+   */
 
     auto predict_fn = [this, name, input_type, policy, latency_slo_micros,
                        app_metrics](
         std::shared_ptr<HttpServer::Response> response,
         std::shared_ptr<HttpServer::Request> request) {
       try {
-        std::vector<std::string> models = get_linked_models_for_app(name);
-        std::vector<VersionedModelId> versioned_models;
-        {
-          std::unique_lock<std::mutex> l(current_model_versions_mutex_);
+        rapidjson::Document d;
+        clipper::json::parse_json(request->content.string(), d);
+        std::vector<VersionedModelId> versioned_models = {};
+        std::vector<std::string> models;
+        std::string version = clipper::json::get_string(d, "version");
+        if (!version.empty()) {
+          models = clipper::redis::get_model_versions(redis_connection_, name);
           for (auto m : models) {
-            auto version = current_model_versions_.find(m);
-            if (version != current_model_versions_.end()) {
-              versioned_models.emplace_back(m, version->second);
+            if (m.compare(version) == 0) {
+              versioned_models = {clipper::VersionedModelId(name, m)};
+              break;
+            }
+            if (versioned_models.empty()) {
+              throw version_id_error("Requested model does not exist."); 
             }
           }
-        }
+        } else {
+          models = get_linked_models_for_app(name);
+          {
+            std::unique_lock<std::mutex> (current_model_versions_mutex_);
```

Comment:
This syntax isn't quite right. It should be 
```cpp
std::unique_lock<std::mutex> l(current_model_versions_mutex_);
```

Note the **l** before `(current_model_versions_mutex)`